sdk/identity/Azure.Identity/src/OnBehalfOfCredential.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.Identity.Client;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Enables authentication to Azure Active Directory using an On-Behalf-Of flow.``
+    /// </summary>
+    public class OnBehalfOfCredential : TokenCredential
+    {
+        internal readonly MsalConfidentialClient _client;
+        private readonly string _tenantId;
+        private readonly CredentialPipeline _pipeline;
+        private readonly bool _allowMultiTenantAuthentication;
+        private readonly string _clientId;
+        private readonly string _clientSecret;
+        private readonly UserAssertion _userAssertion;
+
+        /// <summary>
+        /// Protected constructor for mocking.
+        /// </summary>
+        protected OnBehalfOfCredential()
+        { }
+
+        /// <summary>
+        /// Creates an instance of the OnBehalfOfCredential with the details needed to authenticate against Azure Active Directory with the specified certificate path
+        /// (via the <see cref="OnBehalfOfCredentialOptions"/>).
+        /// </summary>
+        /// <param name="tenantId">The Azure Active Directory tenant (directory) Id of the service principal.</param>
+        /// <param name="clientId">The client (application) ID of the service principal</param>
+        /// <param name="userAssertion">The access token that will be used by <see cref="OnBehalfOfCredential"/> as the user assertion when requesting On-Behalf-Of tokens.</param>
+        /// <param name="options">Options that allow to configure the management of the requests sent to the Azure Active Directory service.</param>
+        public OnBehalfOfCredential(string tenantId, string clientId, string userAssertion, OnBehalfOfCredentialOptions options)
```

Comment:
I don't like how this constructor relies on `OnBehalfOfCredentialOptions.CertificatePath` being set. Without carefully reading the documentation, this detail is easy to miss. I'm leaning towards removing `CertificatePath` support from this credential as it introduces a lot of edge cases, and we can add support if it's requested later. What do you think?

Reply:
I'm fine with removing it, but we do throw with an appropriate error if the option is missing.